### PR TITLE
[ready to deploy] - Markers locked to P5 stimuli logged into Raw csv file

### PR DESCRIPTION
**Pull request by kylemath**
Worked example of how markers from p5 sketch get put into window on a given draw frame, then they get logged into the csv file asap 

This saves csv files with a column of 0 and 1 (could be any string or number) indicating the conditions, 1's are on for multiple rows, each row is a single time point now (interval and duration are both = 1) so there are no charts possible either. 

Screen Shot 2020-01-05 at 2 07 57 AM

This seems to work so i will make some new sketches tomorrow with p3 task of rare and frequency stimuli.

### CODE REVIEW

File: src/components/PageSwitcher/PageSwitcher.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
- import React, { useState, useCallback } from "react";
+import React, { useState, useCallback } from "react";
```
Comment (kylemath) [inline src/components/PageSwitcher/PageSwitcher.js]: nothing new here

File: src/components/PageSwitcher/components/EEGEduAnimate/EEGEduAnimate.js [hunk 2]
```diff
@@ -36,7 +36,8 @@ export function getSettings () {
     interval: 16,
     bins: 256,
     duration: 128,
-    srate: 256
+    srate: 256,
+    name: 'Animate'
   }
```
Comment (kylemath) [inline src/components/PageSwitcher/components/EEGEduAnimate/EEGEduAnimate.js]: bug fix

File: src/components/PageSwitcher/components/EEGEduEvoked/EEGEduEvoked.js [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+  // take one sample from selected observable object for headers
+  localObservable$ = window.multicastEvoked$.pipe(
+  take(1)
+  );
+  //take one sample to get header info
+  localObservable$.subscribe({ 
+  next(x) { 
+    dataToSave.push(
+      "Timestamp (ms),",
+      "Marker,",
```
Comment (kylemath) [inline src/components/PageSwitcher/components/EEGEduEvoked/EEGEduEvoked.js]: extra header

File: src/components/PageSwitcher/components/EEGEduEvoked/EEGEduEvoked.js [hunk 4]
```diff
@@ -0,0 +1,280 @@
+import React from "react";
```
Comment (kylemath) [inline src/components/PageSwitcher/components/EEGEduEvoked/EEGEduEvoked.js]: this is just the raw module copied, with a p5 flashing ellipse sketch in the modal window, only two other lines are different below

File: src/components/PageSwitcher/components/EEGEduEvoked/EEGEduEvoked.js [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+  });
+  // put selected observable object into local and start taking samples
+  localObservable$ = window.multicastEvoked$.pipe(
+  take(numSamplesToSave)
+  );
+
+  // now with header in place subscribe to each epoch and log it
+  localObservable$.subscribe({
+    next(x) { 
+      dataToSave.push(Date.now() + "," + window.marker + "," + Object.values(x).join(",") + "\n");
```
Comment (kylemath) [inline src/components/PageSwitcher/components/EEGEduEvoked/EEGEduEvoked.js]: this adds a column for window.marker, which is assigned false above to initialize (making it actually 3 lines changed), and changed inside the sketch

File: src/components/PageSwitcher/components/EEGEduEvoked/sketchEvoked.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  p.mousePressed = function () {
+    p.background(256);
+  }
+
+  p.draw = function () {
+    p.background(255);
+    x = x+1;
+    if (x % 11 === 0) {
+      p.fill(0, 0, 0);  
+      window.marker = 1;
```
Comment (kylemath) [inline src/components/PageSwitcher/components/EEGEduEvoked/sketchEvoked.js]: just do this whenever you want to update what is in that column

File: src/components/PageSwitcher/components/EEGEduEvoked/sketchEvoked.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    } else {
+      p.fill(255, 255, 255);
+      window.marker = 0;
```
Comment (kylemath) [inline src/components/PageSwitcher/components/EEGEduEvoked/sketchEvoked.js]: or this if no markers

File: src/components/PageSwitcher/components/EEGEduEvoked/translations/en.json [hunk 8]
```diff
@@ -0,0 +1,12 @@
+{
+  "title": "Stimulus Evoked Event-related potential (ERP)",
+  "description": [
+  	"The electrical activity evoked by individual presentations of stimuli does create small changes in electrical voltage. ", 
+  	"These changes are, however, very small, about 1-10 microVolts (\u03BCV), and the noise from other things like eye movements and muscles is sometimes 10x as large. ",
+  	"This noise can be mitigated by averaging this evoked activity over repeated presentations of the same stimulus. ",
+  	"This average voltage in response to a stimulus is called an Event-related potential (ERP) or evoked potential (EP). ",
+  	"Here..."
```
Comment (kylemath) [inline src/components/PageSwitcher/components/EEGEduEvoked/translations/en.json]: will flesh out later when more sane

File: src/components/PageSwitcher/components/EEGEduHeartSpectra/translations/en.json [hunk 9]
```diff
@@ -8,6 +8,6 @@
     "Along the horizontal axis is the beats per minute, or the frequency of the peaks in your ECG. ", 
     "The vertical y-axis shows the power of the rhythms in the data at each frequency, or how large the changes are between peak and through of the oscillations. "
   ],
-  "xlabel": "Frequency (Hz)",
+  "xlabel": "Heart Frequency (BPM)",
```
Comment (kylemath) [inline src/components/PageSwitcher/components/EEGEduHeartSpectra/translations/en.json]: bug fix

**Comment by kylemath** [comment]
since going to be would be crazy, I tested the markers a bit in xcell

Screen Shot 2020-01-05 at 2 25 26 AM
These are the differences between adjacent time stamps in ms, the period of sampling, and what we expect if it is working exactly at 256 Hz. We are a bit off, pretty good, and some longer delays

period | 3.90625 | expected
period | 3.175087544 | estimated
srate | 256 | expected
srate | 314.9519458 | estimated

---
I also checked the difference between the onset of each stimulus, since the p5 sketch is attempting to use the same number of cycles in each they should be the same (this is not a good test  since that may not be true)

I got these periods between stimuli in ms:


434
432
426
424
422
421
410
407
406
404
402
400
400
399

average = 413.4 ms, and range < 40 ms

we need to make sure the stimuli are actually spaced evenly or this test could have two influences, the delay in the markers and the jitter in the timing of p5

all in all

VERY PROMISING

**Comment by korymath** [comment]
10 minutes. 

Just the gall of adding an annotated screenshot of the timestamps is delightful.

**Comment by kylemath** [comment]
got this working well, on both laptop and phone, shown series of circles, press button or tap when red, saves data with markers for stimulus type, and for responses

**Comment by kylemath** [comment]
dataExample
saves data like this with either touchscreen or spacebar markers for responses

File: src/components/PageSwitcher/components/EEGEduEvoked/EEGEduEvoked.js [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+  take(1)
+  );
+  //take one sample to get header info
+  localObservable$.subscribe({ 
+  next(x) { 
+    dataToSave.push(
+      "Timestamp (ms),",
+      "Marker,",
+      "SpaceBar,",
+      "TouchMarker,",
```
Comment (kylemath) [inline src/components/PageSwitcher/components/EEGEduEvoked/EEGEduEvoked.js]: two extra columns for responses

File: src/components/PageSwitcher/components/EEGEduEvoked/sketchEvoked.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    p.frameRate(30);
+    p.noStroke();
+
+  };
+
+  p.windowResized = function() {
+    p.createCanvas(300, 300);
+  }
+
+  p.touchStarted = function() {
```
Comment (kylemath) [inline src/components/PageSwitcher/components/EEGEduEvoked/sketchEvoked.js]: if touchscreen tapped put marker

File: src/components/PageSwitcher/components/EEGEduEvoked/sketchEvoked.js [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+  if (window.touchMarker === 0) {
+    window.touchMarker = 255;
+  } else {
+    window.touchMarker = 0;
+  }
+}
+
+  p.draw = function () {
+
+    if (p.keyIsPressed === true) {
```
Comment (kylemath) [inline src/components/PageSwitcher/components/EEGEduEvoked/sketchEvoked.js]: if spacebar pressed put marker to window

File: src/components/PageSwitcher/components/EEGEduEvoked/sketchEvoked.js [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+      window.responseMarker = p.keyCode;
+    } else {
+      window.responseMarker = 0;
+    }
+    p.background(255);
+    x = x+1;
+    // var num = int(random(0, 11));
+    if (x % 20 === 0) { // When a target shown (every ith frame for now)
```
Comment (kylemath) [inline src/components/PageSwitcher/components/EEGEduEvoked/sketchEvoked.js]: for now not random space between, that is ok

File: src/components/PageSwitcher/components/EEGEduEvoked/sketchEvoked.js [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+ let x = 0;
+ let thisRand = 0.5; //for random choice of target type
+ let targProp = 0.25;
+ let isTarget = false;
+ 
+
+ p.setup = function () {
+    p.createCanvas(300, 300);
+    p.frameRate(30);
```
Comment (kylemath) [inline src/components/PageSwitcher/components/EEGEduEvoked/sketchEvoked.js]: added this to this and the ssvep animations to aim for lowest common denominator of hardware people may have, and set this common amongst them.

File: src/components/PageSwitcher/components/EEGEduEvoked/EEGEduEvoked.js [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+    </Card>
+  )
+}
+
+export function renderRecord(recordPopChange, recordPop, status, Settings) {
+  return (
+    <Card title={'Run ERP experiment'} sectioned>
+      <Card.Section>
+        <p>
+          {"Clicking this button will begin the experiment so check your data quality on the raw module first. "}
```
Comment (kylemath) [inline src/components/PageSwitcher/components/EEGEduEvoked/EEGEduEvoked.js]: lots of explaination here for what will happen, and an example of what data will look like, we should add this this semester for the other modules as well

**Pull request merged**